x-pack/plugins/apm/server/lib/alerts/register_error_count_alert_type.ts

```diff
@@ -102,8 +110,13 @@ export function registerErrorCountAlertType({
         const alertInstance = services.alertInstanceFactory(
           AlertType.ErrorCount
         );
+
         alertInstance.scheduleActions(alertTypeConfig.defaultActionGroupId, {
-          serviceName: alertParams.serviceName,
+          serviceName:
+            alertParams.serviceName ||
+            getCommaSeparetedAggregationKey(
+              response.aggregations?.services.buckets
+            ),
```

Comment:
I don't think we should comma separate them if we have iterators and can create a default message

```suggestion
          serviceName: alertParams.serviceName
          services: services.buckets
```

Reply:
Thinking about this some more: since global alerts is not a highly requested feature but something we are trying out, I think we should scope it to the simplest implementation possible.

It looks like the implementation would be simpler if we simpler if we do as you first suggested: sending out separate alerts for each service.

If needed we can always improve on this and batch alerts together. But for now I think it's better to keep it simple until we know what is needed (and all this might change in the future anyway)

Reply:
that sounds good to me too. But what about when environment and transaction type are set to `any` too? I don't think creating one alert per service-environment-transaction_type is a good idea. We'd end up with a lot of combinations.

Maybe one alert per service, and then aggregate on environment and transaction type?!

Reply:
> Maybe one alert per service, and then aggregate on environment and transaction type?!

Then we are back to batching, right? In that case we might as well also batch send the services.

Reply:
I am not even sure if we can create multiple alerts at once. Since we use the Alerting flyout, which is provided by the alerting team.

Reply:
> I am not even sure if we can create multiple alerts at once. Since we use the Alerting flyout, which is provided by the alerting team.

This doesn't require any ui changes. It's a matter of sending out multiple alerts:
```ts
services.forEach((serviceName) => {
  transactionTypes.forEach((transactionType) => {
    environments.forEach((environment) => {
        alertInstance.scheduleActions(alertTypeConfig.defaultActionGroupId, {
          serviceName,
          transactionType,
          environment
        });
     });
})
```

Reply:
ah ok, I had in mind that we wanted to create multiple alerts itself, not actions.... but that is what we want...

Reply:
> ah ok, I had in mind that we wanted to create multiple alerts itself, not actions.... but that is what we want...

Btw. I'm not an alerting expert so above might be wrong. I thought that was how to invoke an alert multiple times.

Reply:
it's almost that, we also need to create the alert instance inside the loop `services.alertInstanceFactory`

Reply:
My understanding is as follows: the alert type definition is called the Alert Type. The alert that you create via the API, with a set of parameters, is called the Alert. The alert instance you create server-side (via `alertInstanceFactory(id)`) when you want to schedule actions, is called the Alert Instance. Actions can only be scheduled once per alert instance. So if you want to fire multiple actions, you'll have to create multiple alert instances (with a unique id). We probably should check with the alerting team how we should handle this. I think Uptime has a similar alert.

Reply:
I am following what is described in the documentation https://github.com/elastic/kibana/tree/master/x-pack/plugins/alerts#example

Reply:
Aren't you saying the same thing as @cauemarcondes ? That we need to create a new instance per action we want to fire:

```ts
services.forEach((serviceName) => {
  transactionTypes.forEach((transactionType) => {
    environments.forEach((environment) => {
        const alertInstance = services.alertInstanceFactory(AlertType.ErrorCount);
        alertInstance.scheduleActions(alertTypeConfig.defaultActionGroupId, {
          serviceName,
          transactionType,
          environment
        });
     });
})
```

Reply:
@sqren yes, I wrote my comment without seeing Cauê's. One comment on your example: I think the id needs to be unique (per executor).